File: src/games/sudoku/gameplay/analysis/techniqueModules/xyWing.ts
import { popcount } from '../bitmask';
import { arePeerCells, buildCandidateEliminationMove, buildPlacementHouseRefs, getCellCandidatesByIndex, isSameCell } from '../techniqueHelpers';
import type { SudokuTechniqueDispatcher } from '../techniqueModuleTypes';

export const xyWingTechnique: SudokuTechniqueDispatcher = {
  technique: 'xy-wing',
  tier: 'hard',
  findMove(state) {
    const bivalueCells = state.board
      .map((value, index) => ({ value, index }))
      .filter(({ value, index }) => value === 0 && popcount[state.candidateMask[index]] === 2)
      .map(({ index }) => ({
        row: Math.floor(index / 9),
        col: index % 9,
        index,
        candidates: getCellCandidatesByIndex(state, index),
      }));

    for (const pivot of bivalueCells) {
      const wings = bivalueCells.filter((cell) => !isSameCell(cell, pivot) && arePeerCells(cell, pivot));

      for (const leftWing of wings) {
        for (const rightWing of wings) {
          if (isSameCell(leftWing, rightWing)) {
            continue;
          }

          const leftSharedDigits = pivot.candidates.filter((digit) => leftWing.candidates.includes(digit));
          const rightSharedDigits = pivot.candidates.filter((digit) => rightWing.candidates.includes(digit));
          if (leftSharedDigits.length !== 1 || rightSharedDigits.length !== 1) {
            continue;
          }
          if (leftSharedDigits[0] === rightSharedDigits[0]) {
            continue;
          }

          const leftExtraDigits = leftWing.candidates.filter((digit) => !pivot.candidates.includes(digit));
          const rightExtraDigits = rightWing.candidates.filter((digit) => !pivot.candidates.includes(digit));
          if (leftExtraDigits.length !== 1 || rightExtraDigits.length !== 1) {
            continue;
          }

          const zDigit = leftExtraDigits[0];
          if (zDigit !== rightExtraDigits[0]) {
            continue;
          }

          const move = buildCandidateEliminationMove({
            technique: 'xy-wing',
            eliminations: state.board
              .map((value, index) => ({ value, index }))
              .filter(({ value }) => value === 0)
              .filter(({ index }) => ![pivot.index, leftWing.index, rightWing.index].includes(index))
              .filter(({ index }) => {
                const row = Math.floor(index / 9);
                const col = index % 9;
                return arePeerCells({ row, col }, leftWing) && arePeerCells({ row, col }, rightWing);
              })
              .filter(({ index }) => getCellCandidatesByIndex(state, index).includes(zDigit))
              .map(({ index }) => ({ index, digit: zDigit })),
            evidenceCells: [pivot.index, leftWing.index, rightWing.index],
            houses: buildPlacementHouseRefs(pivot.row, pivot.col),
          });
          if (move) {
            return move;
          }
        }
      }
    }

    return null;
  },
};

File: engine/src/main/kotlin/com/quietgrid/engine/sudoku/techniques/XyWing.kt
package com.quietgrid.engine.sudoku.techniques

import com.quietgrid.engine.core.Difficulty
import com.quietgrid.engine.sudoku.*

object XyWingTechnique : SudokuTechniqueDispatcher {
    override val technique = SudokuTechnique.XY_WING
    override val tier = Difficulty.HARD

    private data class BivalueCell(val row: Int, val col: Int, val index: Int, val candidates: List<Int>)

    override fun findMove(state: SudokuBitmaskState): SudokuCanonicalMove? {
        val bivalueCells = state.board.indices
            .filter { state.board[it] == 0 && popcount[state.candidateMask[it]] == 2 }
            .map { BivalueCell(it / 9, it % 9, it, getCellCandidatesByIndex(state, it)) }
        val complexity = bivalueCells.size

        for (pivot in bivalueCells) {
            val wings = bivalueCells.filter { it.index != pivot.index && arePeerIndexes(it.index, pivot.index) }

            for (leftWing in wings) {
                for (rightWing in wings) {
                    if (leftWing.index == rightWing.index) continue
                    val leftShared = pivot.candidates.filter { leftWing.candidates.contains(it) }
                    val rightShared = pivot.candidates.filter { rightWing.candidates.contains(it) }
                    if (leftShared.size != 1 || rightShared.size != 1 || leftShared[0] == rightShared[0]) continue

                    val leftExtra = leftWing.candidates.filter { !pivot.candidates.contains(it) }
                    val rightExtra = rightWing.candidates.filter { !pivot.candidates.contains(it) }
                    if (leftExtra.size != 1 || rightExtra.size != 1) continue
                    val zDigit = leftExtra[0]
                    if (zDigit != rightExtra[0]) continue

                    val eliminations = state.board.indices
                        .filter { state.board[it] == 0 && it != pivot.index && it != leftWing.index && it != rightWing.index }
                        .filter { index ->
                            val cell = SudokuCellRef(index / 9, index % 9)
                            arePeerCells(cell, SudokuCellRef(leftWing.row, leftWing.col)) && arePeerCells(cell, SudokuCellRef(rightWing.row, rightWing.col))
                        }
                        .filter { getCellCandidatesByIndex(state, it).contains(zDigit) }
                        .map { it to zDigit }

                    val move = buildCandidateEliminationMove(
                        SudokuTechnique.XY_WING, eliminations,
                        listOf(pivot.index, leftWing.index, rightWing.index),
                        buildPlacementHouseRefs(pivot.row, pivot.col), complexity,
                    )
                    if (move != null) return move
                }
            }
        }
        return null
    }
}
